zebROS_ws/src/controllers_2023/src/elevator_controller.cpp

```diff
@@ -0,0 +1,387 @@
+		}
+		const ElevatorCommand_2023 setpoint = *(position_command_.readFromRT());
+		elevator_joint_.setCommand(setpoint.GetPosition());
+
+		//if we're not climbing, add an arbitrary feed forward to hold the elevator up
+		if(!setpoint.GetGoSlow())
+		{
+			elevator_joint_.setMotionAcceleration(config_.motion_magic_acceleration);
+			elevator_joint_.setMotionCruiseVelocity(config_.motion_magic_velocity);
+			elevator_joint_.setPIDFSlot(0);
+			// Add arbitrary feed forward for upwards motion
+			// We could have arb ff for both up and down, but seems
+			// easier (and good enough) to tune PID for down motion
+			// and add an arb FF correction for up
+
+			if(elevator_joint_.getPosition() >= config_.stage_2_height && last_position_ <= config_.stage_2_height) {
+				elevator_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+				elevator_joint_.setDemand1Value(config_.arb_feed_forward_high);
+			}
+			else if (elevator_joint_.getPosition() <= config_.stage_2_height && last_position_ >= config_.stage_2_height) {
+				elevator_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+				elevator_joint_.setDemand1Value(config_.arb_feed_forward_low);
+			}
+
+
+			//for now, up and down PID is the same, so slot 1 is used for climbing
+			/*
+			if(last_setpoint_ != setpoint) {
+				if(setpoint > elevator_joint_.getPosition()) {
+					elevator_joint_.setPIDFSlot(0);
+				}
+				else {
+					elevator_joint_.setPIDFSlot(1);
+				}
+			}
+			last_setpoint_ = setpoint;
+			*/
+		}
+		
+	}
+	else
+	{
+		elevator_joint_.setMode(hardware_interface::TalonMode_PercentOutput);
+		if ((ros::Time::now() - last_time_down_).toSec() < config_.elevator_zeroing_timeout)
+		{
+			// Not yet zeroed. Run the elevator down slowly until the limit switch is set.
+			ROS_INFO_STREAM_THROTTLE(0.25, "Zeroing elevator with percent output: "
+									 << config_.elevator_zeroing_percent_output);
+			elevator_joint_.setCommand(config_.elevator_zeroing_percent_output);
+		}
+		//check stream
+		else
+		{
+			// Stop moving to prevent motor from burning out
+			ROS_INFO_STREAM_THROTTLE(0.25, "Elevator timed out");
+			elevator_joint_.setCommand(0);
+		}
+
+		// If not zeroed but enabled, check if the arm is moving down
+		if ((elevator_joint_.getMode() == hardware_interface::TalonMode_Disabled) ||
+			(elevator_joint_.getSpeed() < 0)) // TODO : param
+		{
+			// If moving down, or disabled and thus not expected to move down, reset the timer
+			last_time_down_ = ros::Time::now();
+		}
+	}
+	last_position_ = elevator_joint_.getPosition();
+	last_mode_ = elevator_joint_.getMode();
+}
+
+void ElevatorController_2023::stopping(const ros::Time &/*time*/)
+{
+}
+
+//Command Service Function
+bool ElevatorController_2023::cmdService(controllers_2023_msgs::ElevatorSrv::Request  &req,
+									controllers_2023_msgs::ElevatorSrv::Response &/*response*/)
+{
+	if (req.position > 1.7) // TODO : get real measurement, make a param
```

Comment:
The TODO here is worth doing
 - make a member variable to hold this max elevator height variable
 - in init, add an nh.getParam call to load it from the config file
 - add an entry in the yaml config file for this new variable.  You can default to 1.7 which lets us test on the 2019 elevator, we'll change it later when we have the actual 2023 hardware